openedx/core/djangoapps/content/search/api.py

```diff
@@ -0,0 +1,486 @@
+    status_cb("Checking index...")
+    with _index_rebuild_lock() as temp_index_name:
+        if _index_exists(temp_index_name):
+            status_cb("Temporary index already exists. Deleting it...")
+            _wait_for_meili_task(client.delete_index(temp_index_name))
+
+        status_cb("Creating new index...")
+        _wait_for_meili_task(
+            client.create_index(temp_index_name, {'primaryKey': 'id'})
+        )
+        new_index_created = client.get_index(temp_index_name).created_at
+
+        yield temp_index_name
+
+        if not _index_exists(INDEX_NAME):
+            # We have to create the "target" index before we can successfully swap the new one into it:
+            status_cb("Preparing to swap into index (first time)...")
+            _wait_for_meili_task(client.create_index(INDEX_NAME))
+        status_cb("Swapping index...")
+        client.swap_indexes([{'indexes': [temp_index_name, INDEX_NAME]}])
+        # If we're using an API key that's restricted to certain index prefix(es), we won't be able to get the status
+        # of this request unfortunately. https://github.com/meilisearch/meilisearch/issues/4103
+        while True:
+            time.sleep(1)
+            if client.get_index(INDEX_NAME).created_at != new_index_created:
+                status_cb("Waiting for swap completion...")
+            else:
+                break
+        status_cb("Deleting old index...")
+        _wait_for_meili_task(client.delete_index(temp_index_name))
+
+
+def _recurse_children(block, fn, status_cb: Callable[[str], None] | None = None) -> None:
+    """
+    Recurse the children of an XBlock and call the given function for each
+
+    The main purpose of this is just to wrap the loading of each child in
+    try...except. Otherwise block.get_children() would do what we need.
+    """
+    if block.has_children:
+        for child_id in block.children:
+            try:
+                child = block.get_child(child_id)
+            except Exception as err:  # pylint: disable=broad-except
+                log.exception(err)
+                if status_cb is not None:
+                    status_cb(f"Unable to load block {child_id}")
+            else:
+                fn(child)
+
+
+def only_if_meilisearch_enabled(f):
+    """
+    Only call `f` if meilisearch is enabled
+    """
+    @wraps(f)
+    def wrapper(*args, **kwargs):
+        """Wraps the decorated function."""
+        if is_meilisearch_enabled():
+            return f(*args, **kwargs)
+    return wrapper
+
+
+def is_meilisearch_enabled() -> bool:
+    """
+    Returns whether Meilisearch is enabled
+    """
+    if hasattr(settings, "MEILISEARCH_ENABLED"):
+        return settings.MEILISEARCH_ENABLED
+
+    return False
+
+
+def rebuild_index(status_cb: Callable[[str], None] | None = None) -> None:
+    """
+    Rebuild the Meilisearch index from scratch
+    """
+    def nop(_message):
+        pass
+
+    if status_cb is None:
+        status_cb = nop
```

Comment:
If no `status_cb` is passed, we should log using the `log` function from line 35.

Reply:
Done: 814d0a6680c02789ddf2d8d0b8d42bc5f693ea3f